impl/src/main/java/org/jboss/weld/bootstrap/WeldUnusedMetadataExtension.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2018, Red Hat, Inc. and/or its affiliates, and individual
+ * contributors by the @authors tag. See the copyright.txt in the
+ * distribution for a full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.bootstrap;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Initialized;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.Instance;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.enterprise.inject.spi.ProcessInjectionPoint;
+
+import org.jboss.weld.manager.BeanManagerImpl;
+import org.jboss.weld.resolution.Resolvable;
+import org.jboss.weld.resolution.ResolvableBuilder;
+import org.jboss.weld.util.reflection.Reflections;
+
+/**
+ * This optional extension collects all injection points
+ * <ul>
+ * <li>of Java EE components,</li>
+ * <li>with {@link Instance} required type,</li>
+ * </ul>
+ * so that Weld is able to identify unused beans better.
+ *
+ * @author Martin Kouba
+ */
+public class WeldUnusedMetadataExtension implements Extension {
+
+    private Set<InjectionPoint> componentInjectionPoints;
+
+    private Set<InjectionPoint> instanceInjectionPoints;
+
+    WeldUnusedMetadataExtension() {
+        this.componentInjectionPoints = new HashSet<>();
+        this.instanceInjectionPoints = new HashSet<>();
+    }
+
+    void processInjectionPoints(@Observes ProcessInjectionPoint<?, ?> event) {
```

Comment:
Any reason for this being package protected? Same for `clear()` method.

Reply:
There is no need to have them public, right? `WeldUnusedMetadataExtension` itself must be public but we don't want to "tempt" passersby ;-) 

I think it's always good to minimize visibility.

Reply:
Yes, I was thinking they could just be private then.